test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +53,475 @@ var (
+				// still match other possible ops.
+				firstError = err2
+			}
+		}
+	}
+	return fmt.Errorf("cannot unmarshal %s into any known op type: %v", string(b), firstError)
+}
+
+// realOp is an interface that is implemented by different structs. To evaluate
+// the validity of ops at parse-time, a isValid function must be implemented.
+type realOp interface {
+	// isValid verifies the validity of the op args such as node/pod count.
+	isValid() error
+	// collectsMetrics checks if the op collects metrics.
+	collectsMetrics() bool
+}
+
+// createNodesOp defines an op where nodes are created as a part of a workload.
+type createNodesOp struct {
+	// Number of nodes to create.
+	CreateNodes int
+	// Path to spec file describing the nodes to create. Optional.
 	NodeTemplatePath *string
-	// At most one of the following strategies can be defined. If not specified, default to TrivialNodePrepareStrategy.
+	// At most one of the following strategies can be defined. Optional, defaults
+	// to TrivialNodePrepareStrategy if unspecified.
 	NodeAllocatableStrategy  *testutils.NodeAllocatableStrategy
 	LabelNodePrepareStrategy *testutils.LabelNodePrepareStrategy
 	UniqueNodeLabelStrategy  *testutils.UniqueNodeLabelStrategy
 }
 
-type podCase struct {
-	Num                               int
-	PodTemplatePath                   *string
+func (cno *createNodesOp) isValid() error {
+	if cno.CreateNodes <= 0 {
+		return fmt.Errorf("number of nodes cannot be non-positive")
+	}
+	return nil
+}
+
+func (*createNodesOp) collectsMetrics() bool {
+	return false
+}
+
+// createPodsOp defines an op where pods are scheduled as a part of a workload.
+// The test can block on the completion of this op before moving forward or
+// continue asynchronously.
+type createPodsOp struct {
+	// Number of pods to schedule.
+	CreatePods int
+	// Whether or not to enable metrics collection for this createPodsOp.
+	// Optional. Both CollectMetrics and SkipWaitToCompletion cannot be true at
+	// the same time for a particular createPodsOp.
+	CollectMetrics bool
+	// Namespace the pods should be created in. Optional, defaults to a unique
+	// namespace of the format "namespace-<number>".
+	Namespace *string
+	// Path to spec file describing the pods to schedule. Optional.
+	PodTemplatePath *string
+	// Whether or not to wait for all pods in this op to get scheduled. Optional,
+	// defaults to false.
+	SkipWaitToCompletion bool
+	// Persistent volume settings for the pods to be scheduled. Optional.
 	PersistentVolumeTemplatePath      *string
 	PersistentVolumeClaimTemplatePath *string
 }
 
-// simpleTestCases defines a set of test cases that share the same template (node spec, pod spec, etc)
-// with testParams(e.g., NumNodes) being overridden. This provides a convenient way to define multiple tests
-// with various sizes.
-type simpleTestCases struct {
-	Template testCase
-	Params   []testParams
+func (cpo *createPodsOp) isValid() error {
```

Comment:
should we validate that `CollectMetrics` and `SkipWaitToCompletion` can't be true at the same time?

Reply:
I did that originally, but then I added support for them both to be true. The metrics collector will run in the background when the benchmark finishes automatically.